Potential issue with RNG in `gcp_opt` testing

Prior to 573dcd9 (PR #450), the tests for `gcp_opt` were working. After the introduction of a few new tests to support the new capabilities covered in that PR, the tests were failing in a non-deterministic manner. A temporary fix was to increase all of the `max_iters` values to 3 for all `gcp_opt` solvers in `tests/gcp/test_optimizers.py`, so that there were enough iterations for the solver to converge to the expected solution.

I suspect that the state of the RNG was changed with the introduction of the new `gcp_opt` tests, propagating to the failures. Using the pre-PR #450 version [here](https://github.com/sandialabs/pyttb/blob/c6d5fe2639d66ffa85ff8c38aac569220d287534/tests/gcp/test_optimizers.py), I can reproduce the errors using successive runs of just this test many times. Below are some of the results, showing passing, failing at location A, and failing at location B.

Passing:
```commandline
$ pytest tests -k "test_optimizers"
=============================================================================================================== test session starts ===============================================================================================================
platform darwin -- Python 3.12.9, pytest-8.4.0, pluggy-1.6.0
rootdir: <>
configfile: pyproject.toml
plugins: cov-6.1.1
collected 608 items / 603 deselected / 5 selected                                                                                                                                                                                                 

tests/gcp/test_optimizers.py .....                                                                                                                                                                                                          [100%]

======================================================================================================== 5 passed, 603 deselected in 0.18s ========================================================================================================

```

Failing at line 73:

```commandline
$ pytest tests -k "test_optimizers"
=============================================================================================================== test session starts ===============================================================================================================
platform darwin -- Python 3.12.9, pytest-8.4.0, pluggy-1.6.0
rootdir: <>
configfile: pyproject.toml
plugins: cov-6.1.1
collected 608 items / 603 deselected / 5 selected                                                                                                                                                                                                 

tests/gcp/test_optimizers.py .F...                                                                                                                                                                                                          [100%]

==================================================================================================================== FAILURES =====================================================================================================================
____________________________________________________________________________________________________________________ test_adam ____________________________________________________________________________________________________________________

generate_problem = (tensor of shape (2, 2) with order F
data[:, :] =
[[1. 0.]
 [0. 1.]], ktensor of shape (2, 2) with order F
weights=[1....0] =
[[1. 1.]
 [1. 1.]]
factor_matrices[1] =
[[1. 1.]
 [1. 1.]], <pyttb.gcp.samplers.GCPSampler object at 0x107af3e60>)

    def test_adam(generate_problem):
        dense_data, model, sampler = generate_problem
    
        solver = Adam(max_iters=1, epoch_iters=1)
        result, info = solver.solve(
            model, dense_data, gaussian, gaussian_grad, sampler=sampler
        )
        assert isinstance(info, dict)
        assert (model.full() - dense_data).norm() > (result.full() - dense_data).norm()
    
        # Force bad step
        solver = Adam(max_iters=1, epoch_iters=1)
        result, info = solver.solve(
            model, dense_data, diverging_function_handle, gaussian_grad, sampler=sampler
        )
>       assert model.isequal(result)
E       assert False
E        +  where False = isequal(ktensor of shape (2, 2) with order F\nweights=[1. 1.]\nfactor_matrices[0] =\n[[0.999000000001  0.999000000001 ]\n [0.99900...9990000000025]]\nfactor_matrices[1] =\n[[0.9990000000016667 0.9990000000016667]\n [0.99900000000125   0.99900000000125  ]])
E        +    where isequal = ktensor of shape (2, 2) with order F\nweights=[1. 1.]\nfactor_matrices[0] =\n[[1. 1.]\n [1. 1.]]\nfactor_matrices[1] =\n[[1. 1.]\n [1. 1.]].isequal

tests/gcp/test_optimizers.py:73: AssertionError
============================================================================================================= short test summary info =============================================================================================================
FAILED tests/gcp/test_optimizers.py::test_adam - assert False
=================================================================================================== 1 failed, 4 passed, 603 deselected in 0.20s ===================================================================================================

```

Failing at line 94:

```commandline
$ pytest tests -k "test_optimizers"
=============================================================================================================== test session starts ===============================================================================================================
platform darwin -- Python 3.12.9, pytest-8.4.0, pluggy-1.6.0
rootdir: <>
configfile: pyproject.toml
plugins: cov-6.1.1
collected 608 items / 603 deselected / 5 selected                                                                                                                                                                                                 

tests/gcp/test_optimizers.py ..F..                                                                                                                                                                                                          [100%]

==================================================================================================================== FAILURES =====================================================================================================================
__________________________________________________________________________________________________________________ test_adagrad ___________________________________________________________________________________________________________________

generate_problem = (tensor of shape (2, 2) with order F
data[:, :] =
[[1. 0.]
 [0. 1.]], ktensor of shape (2, 2) with order F
weights=[1....0] =
[[1. 1.]
 [1. 1.]]
factor_matrices[1] =
[[1. 1.]
 [1. 1.]], <pyttb.gcp.samplers.GCPSampler object at 0x10bb5e3c0>)

    def test_adagrad(generate_problem):
        dense_data, model, sampler = generate_problem
    
        solver = Adagrad(max_iters=1, epoch_iters=1)
        result, info = solver.solve(
            model, dense_data, gaussian, gaussian_grad, sampler=sampler
        )
        assert isinstance(info, dict)
        assert (model.full() - dense_data).norm() > (result.full() - dense_data).norm()
    
        # Force bad step
        solver = Adagrad(max_iters=1, epoch_iters=1)
        result, info = solver.solve(
            model, dense_data, diverging_function_handle, gaussian_grad, sampler=sampler
        )
>       assert model.isequal(result)
E       assert False
E        +  where False = isequal(ktensor of shape (2, 2) with order F\nweights=[1. 1.]\nfactor_matrices[0] =\n[[0.9019419324309079 0.9019419324309079]\n [0...7096621545398]]\nfactor_matrices[1] =\n[[0.5097096621545398 0.5097096621545398]\n [0.9019419324309079 0.9019419324309079]])
E        +    where isequal = ktensor of shape (2, 2) with order F\nweights=[1. 1.]\nfactor_matrices[0] =\n[[1. 1.]\n [1. 1.]]\nfactor_matrices[1] =\n[[1. 1.]\n [1. 1.]].isequal

tests/gcp/test_optimizers.py:94: AssertionError
============================================================================================================= short test summary info =============================================================================================================
FAILED tests/gcp/test_optimizers.py::test_adagrad - assert False
=================================================================================================== 1 failed, 4 passed, 603 deselected in 0.20s ===================================================================================================

```

--- a/tests/gcp/test_optimizers.py
+++ b/tests/gcp/test_optimizers.py
@@ -4,6 +4,8 @@
 
 from __future__ import annotations
 
+from typing import TYPE_CHECKING
+
 import numpy as np
 import pytest
 
@@ -12,14 +14,17 @@ from pyttb.gcp import samplers
 from pyttb.gcp.handles import gaussian, gaussian_grad
 from pyttb.gcp.optimizers import LBFGSB, SGD, Adagrad, Adam
 
-global f_est
-f_est = 0.0
+if TYPE_CHECKING:
+    from pyttb.gcp.fg_setup import function_type
 
 
-def diverging_function_handle(data: np.ndarray, model: np.ndarray) -> np.ndarray:
-    global f_est
-    f_est += 1.0
-    return f_est * data
+def diverging_function_handle_factory(function_handle: function_type) -> function_type:
+    """Returns a function that diverges when the gradient is correct."""
+
+    def diverging_function_handle(data: np.ndarray, model: np.ndarray) -> np.ndarray:
+        return -1.0 * function_handle(data, model)
+
+    return diverging_function_handle
 
 
 @pytest.fixture()
@@ -47,9 +52,13 @@ def test_sgd(generate_problem):
         inf_data = np.inf * ttb.tenones(dense_data.shape)
         solver.solve(model, inf_data, gaussian, gaussian_grad, sampler=sampler)
 
-    # Force bad step
+    # Force bad step to check rejection
     result, info = solver.solve(
-        model, dense_data, diverging_function_handle, gaussian_grad, sampler=sampler
+        model,
+        dense_data,
+        diverging_function_handle_factory(gaussian),
+        gaussian_grad,
+        sampler=sampler,
     )
     assert model.isequal(result)
     assert solver._nfails == min(solver._max_iters, solver._max_fails + 1)
@@ -65,10 +74,14 @@ def test_adam(generate_problem):
     assert isinstance(info, dict)
     assert (model.full() - dense_data).norm() > (result.full() - dense_data).norm()
 
-    # Force bad step
+    # Force bad step to check rejection
     solver = Adam(max_iters=3, epoch_iters=1)
     result, info = solver.solve(
-        model, dense_data, diverging_function_handle, gaussian_grad, sampler=sampler
+        model,
+        dense_data,
+        diverging_function_handle_factory(gaussian),
+        gaussian_grad,
+        sampler=sampler,
     )
     assert model.isequal(result)
     assert [np.testing.assert_array_equal(mk, np.zeros_like(mk)) for mk in solver._m]
@@ -86,10 +99,14 @@ def test_adagrad(generate_problem):
     assert isinstance(info, dict)
     assert (model.full() - dense_data).norm() > (result.full() - dense_data).norm()
 
-    # Force bad step
+    # Force bad step to check rejection
     solver = Adagrad(max_iters=3, epoch_iters=1)
     result, info = solver.solve(
-        model, dense_data, diverging_function_handle, gaussian_grad, sampler=sampler
+        model,
+        dense_data,
+        diverging_function_handle_factory(gaussian),
+        gaussian_grad,
+        sampler=sampler,
     )
     assert model.isequal(result)
     assert solver._gnormsum == 0.0

